ui/app/hooks/use-local-storage.ts

```diff
@@ -0,0 +1,79 @@
+import {
+  useCallback,
+  useEffect,
+  useSyncExternalStore,
+  type Dispatch,
+  type SetStateAction,
+} from "react";
+
+type Value = string | null | undefined;
+
+function dispatchStorageEvent(key: string, newValue: Value) {
+  window.dispatchEvent(new StorageEvent("storage", { key, newValue }));
+}
+
+const setLocalStorageItem = (key: string, value: Value) => {
+  const stringifiedValue = JSON.stringify(value);
+  window.localStorage.setItem(key, stringifiedValue);
+  dispatchStorageEvent(key, stringifiedValue);
+};
+
+const removeLocalStorageItem = (key: string) => {
+  window.localStorage.removeItem(key);
+  dispatchStorageEvent(key, null);
+};
+
+const getLocalStorageItem = (key: string) => {
+  return window.localStorage.getItem(key);
+};
+
+const useLocalStorageSubscribe = (callback: (ev: StorageEvent) => void) => {
+  window.addEventListener("storage", callback);
+  return () => window.removeEventListener("storage", callback);
+};
+
+const getLocalStorageServerSnapshot = () => {
+  return null;
+};
+
+export function useLocalStorage<T extends Value>(
+  key: string,
+  initialValue?: T,
+): [T, React.Dispatch<React.SetStateAction<T>>] {
+  const getSnapshot = () => getLocalStorageItem(key);
+
+  const store =
+    useSyncExternalStore(
+      useLocalStorageSubscribe,
+      getSnapshot,
+      getLocalStorageServerSnapshot,
+    ) ?? "";
+
+  const setState = useCallback<Dispatch<SetStateAction<T>>>(
```

Comment:
also generally this file seems pretty complicated for what should be a simple goal. Why's it so complicated?

Reply:
I agree! React v18 brings `useSyncExternalStore` as the official way to sync external state in React, which adds quite a bit of overhead in needing the define more callback functions.

The code in this file is based off of `@uidotdev/usehooks`'s `useLocalStorage()` hook, which handles plenty of edge cases that can crop up in production.